src/expand.rs

```diff
@@ -127,7 +147,7 @@ fn transform_sig(
                 where_token: Default::default(),
                 predicates: Punctuated::new(),
             });
-        for param in &sig.generics.params {
+        for param in sig.generics.params.iter().chain(context.lifetimes()) {
```

Comment:
This causes the impl to mismatch the trait when the impl generics contains more lifetimes.

```rust
#[async_trait]
trait Trait {
    async fn f();
}

#[async_trait]
impl<'a> Trait for &'a () {
    async fn f() {}
}
```

```console
error[E0195]: lifetime parameters or bounds on method `f` do not match the trait declaration
   --> tests/test.rs:315:5
    |
312 |         async fn f();
    |                   - lifetimes in impl do not match this method in trait
...
315 |     #[async_trait]
    |     ^^^^^^^^^^^^^^ lifetimes do not match method in trait
```

You will need to chain only those lifetimes that appear on the impl block's trait, i.e. `impl Trait<'a> for Type<'b>` should chain 'a only.

Reply:
Also, I'd like it if these lifetimes could be filtered down per-function.

```rust
#[async_trait]
trait Trait<'a, 'b> {
    async fn f(_: &'a &'b ()); // chain 'a and 'b
    async fn g(_: &'b ()); // chain 'b only
    async fn h(); // do not chain
}
```

I have a feeling this will reduce the occurrence of lifetime headaches from this change.

Reply:
Done in https://github.com/dtolnay/async-trait/pull/30/commits/8ff8ec2b64a8f1a5429ef405b8895fde33712ff3.